service/history/handler.go

```diff
@@ -703,6 +704,42 @@ func (h *handlerImpl) StartWorkflowExecution(
 	return response, nil
 }
 
+func (h *handlerImpl) DescribeShardDistribution(
+	ctx context.Context,
+	request *types.DescribeShardDistributionRequest,
+) (resp *types.DescribeShardDistributionResponse, retError error) {
+	numShards := h.config.NumberOfShards
+	resp = &types.DescribeShardDistributionResponse{
+		NumberOfShards: 0,
+		Shards:         make(map[int32]string),
+	}
+	if len(*request.Role) == 0 {
+		*request.Role = "cadence-history"
+	}
+
+	offset := request.PageID * request.PageSize
+	// count all the shards but put only the ones falling into the requested page
+	addShardToResponse := func(shardID int, identity string) {
+		if resp.NumberOfShards >= offset && len(resp.Shards) < int(request.PageSize) {
+			resp.Shards[int32(shardID)] = identity
+		}
+		resp.NumberOfShards++
+	}
+
+	for shardID := 0; shardID < numShards; shardID++ {
```

Comment:
Looks like we can just look up shardIDs falling into the request page since the total number of shards will never change?

Reply:
Is this condition true: number_of_shards == any_type_of_shard ? 

If there's a filter we need to count how many shards have this role right?